src/kaleidoscope/hid.cpp

```diff
@@ -150,11 +224,46 @@ uint8_t getKeyboardLEDs() {
 
 
 void sendKeyboardReport() {
+
+  // Before sending the report, we add any modifier flags that are currently allowed,
+  // based on the latest keypress:
+  pressModFlags(current_mod_flags & mod_flags_allowed);
+
+
+  // If a key has just toggled on in this cycle, we might need to send an extra HID report
+  // to the host, because that key might have the same keycode as another key that was
+  // already in the report on the previous cycle. For example, a user could have two
+  // `Key_E` keys in their keymap, in order to avoid repeating the same key with one
+  // finger. Or one might have a `LCTRL(Key_S)` and a plain `Key_S`, and have a reason to
+  // press them in rapid succession. In order to make this work, we need to call
+  // `release()` & `sendReport()` to send a release event to the host so that its normal
+  // repeat-rate-limiting behaviour won't effectively mask the second keypress. Then we
+  // call `press()` to add the keycode back in before sending the normal report.
+  //
+  // In most cases, this won't result in any difference from the previous report (because
+  // the newly-toggled-on keycode won't be in the previous report), so no extra report
+  // will be sent (because we suppress duplicate reports in KeyboardioHID). If there is a
+  // difference in the modifiers byte, an extra report would be sent later, regardless
+  // (also in KeyboardioHID).
```

Comment:
I think this explanation is pretty clear now, but it's still several lines away from some of the code that it applies to. @obra, what do you think?